homeassistant/components/tado/__init__.py

```diff
@@ -43,45 +51,113 @@
 )
 
 
-def setup(hass, config):
-    """Set up of the Tado component."""
-    acc_list = config[DOMAIN]
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Tado component."""
 
-    api_data_list = []
+    hass.data.setdefault(DOMAIN, {})
+
+    if DOMAIN not in config:
+        return True
 
-    for acc in acc_list:
-        username = acc[CONF_USERNAME]
-        password = acc[CONF_PASSWORD]
-        fallback = acc[CONF_FALLBACK]
+    for conf in config[DOMAIN]:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf,
+            )
+        )
 
-        tadoconnector = TadoConnector(hass, username, password, fallback)
-        if not tadoconnector.setup():
-            continue
+    return True
 
-        # Do first update
-        tadoconnector.update()
 
-        api_data_list.append(tadoconnector)
-        # Poll for updates in the background
-        hass.helpers.event.track_time_interval(
-            # we're using here tadoconnector as a parameter of lambda
-            # to capture actual value instead of closuring of latest value
-            lambda now, tc=tadoconnector: tc.update(),
-            SCAN_INTERVAL,
-        )
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Tado from a config entry."""
+
+    _async_import_options_from_data_if_missing(hass, entry)
+
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    fallback = entry.options.get(CONF_FALLBACK, True)
+
+    tadoconnector = TadoConnector(hass, username, password, fallback)
+
+    try:
+        await hass.async_add_executor_job(tadoconnector.setup)
+    except KeyError:
+        _LOGGER.error("Failed to login to tado")
+        return False
+    except RuntimeError as exc:
+        _LOGGER.error("Failed to setup tado: %s", exc)
+        return ConfigEntryNotReady
+    except requests.exceptions.HTTPError as ex:
+        if ex.response.status_code > 400 and ex.response.status_code < 500:
+            _LOGGER.error("Failed to login to tado: %s", ex)
+            return False
+        raise ConfigEntryNotReady
+
+    # Do first update
+    await hass.async_add_executor_job(tadoconnector.update)
+
+    # Poll for updates in the background
+    update_track = async_track_time_interval(
+        hass,
+        # we're using here tadoconnector as a parameter of lambda
+        # to capture actual value instead of closuring of latest value
+        lambda now, tc=tadoconnector: tc.update(),
```

Comment:
This is only necessary inside a for-loop because of how Python works. Since there is no for-loop, you can call it directly
```suggestion
        lambda now: tadoconnector.update(),
```